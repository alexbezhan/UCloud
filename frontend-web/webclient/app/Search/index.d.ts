import {DetailedApplicationSearchReduxState, FullAppInfo} from "Applications";
import {HeaderSearchType} from "DefaultObjects";
import {AdvancedSearchRequest, DetailedFileSearchReduxState, File} from "Files";
import {AdvancedSearchRequest as AppSearchRequest} from "Applications";
import {History} from "history";
import {match} from "react-router-dom";

export interface SearchProps extends SimpleSearchOperations, SimpleSearchStateProps {
    match: match<{priority: string}>;
    /* If extends interface with these, Jest complains */
    history: History;
    location: {
        // TODO There is more here
        search: string
    };
}

export interface SimpleSearchStateProps {
    files: Page<File>;
    filesLoading: boolean;
    applications: Page<FullAppInfo>;
    applicationsLoading: boolean;
    errors: string[];
    search: string;
    fileSearch: DetailedFileSearchReduxState;
    applicationSearch: DetailedApplicationSearchReduxState;
}

export interface SimpleSearchOperations {
    clear: () => void;
    setFilesLoading: (loading: boolean) => void;
    setApplicationsLoading: (loading: boolean) => void;
    searchFiles: (body: AdvancedSearchRequest) => void;
    searchApplications: (body: AppSearchRequest) => void;
    setFilesPage: (page: Page<File>) => void;
    setApplicationsPage: (page: Page<FullAppInfo>) => void;
    setSearch: (search: string) => void;
    setPrioritizedSearch: (st: HeaderSearchType) => void;
    toggleAdvancedSearch: () => void;
    setActivePage: () => void;
    setRefresh: (refresh?: () => void) => void;
}