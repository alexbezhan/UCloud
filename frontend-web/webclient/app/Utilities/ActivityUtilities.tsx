import {ActivityFilter} from "Activity";
import {ScrollRequest} from "Scroll";
import {buildQueryString} from "./URIUtilities";

interface ParamsType extends Omit<ActivityFilter, "maxTimestamp" | "minTimestamp">, Partial<ScrollRequest<number>> {
    maxTimestamp?: number;
    minTimestamp?: number;
}

export const activityQuery = (scroll: ScrollRequest<number>, filter?: ActivityFilter) => {
    const params: ParamsType = {};
    if (!!scroll.offset) params.offset = scroll.offset;
    params.scrollSize = scroll.scrollSize;

    if (filter !== undefined) {
        if (filter.type !== undefined) params.type = filter.type;
        if (filter.collapseAt !== undefined) params.collapseAt = filter.collapseAt;
        if (filter.maxTimestamp !== undefined) params.maxTimestamp = filter.maxTimestamp.getTime();
        if (filter.minTimestamp !== undefined) params.minTimestamp = filter.minTimestamp.getTime();
        if (filter.user !== undefined) params.user = filter.user;
    }

    return buildQueryString("/activity/browse", params);
};

export const activityStreamByPath = (path: string) =>
    `/activity/by-path?path=${encodeURIComponent(path)}`;
