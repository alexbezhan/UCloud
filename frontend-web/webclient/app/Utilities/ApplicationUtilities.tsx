import {
    ApplicationInvocationDescription,
    ApplicationMetadata,
    ApplicationParameter,
    AppState,
    ParameterTypes,
    RunsSortBy
} from "Applications";
import Cloud from "Authentication/lib";
import {SortOrder} from "Files";
import * as React from "react";
import {snackbarStore} from "Snackbar/SnackbarStore";
import {Page} from "Types";
import {addStandardDialog} from "UtilityComponents";
import {errorMessageOrDefault, removeTrailingSlash} from "UtilityFunctions";
import {expandHomeFolder} from "./FileUtilities";

export const hpcJobQueryPost = "/hpc/jobs";

export const hpcJobQuery = (
    id: string,
    stdoutLine: number,
    stderrLine: number,
    stdoutMaxLines: number = 1000,
    stderrMaxLines: number = 1000
) =>
    `/hpc/jobs/follow/${encodeURIComponent(id)}?stdoutLineStart=${stdoutLine}&stdoutMaxLines=${stdoutMaxLines}&stderrLineStart=${stderrLine}&stderrMaxLines=${stderrMaxLines}`;

export function hpcJobsQuery(
    itemsPerPage: number,
    page: number,
    sortOrder?: SortOrder,
    sortBy?: RunsSortBy,
    minTimestamp?: number,
    maxTimestamp?: number,
    filter?: AppState
): string {
    let query = `/hpc/jobs/?itemsPerPage=${itemsPerPage}&page=${page}`;
    if (sortOrder) query = query.concat(`&order=${sortOrder}`);
    if (sortBy) query = query.concat(`&sortBy=${sortBy}`);
    if (minTimestamp != null) query = query.concat(`&minTimestamp=${minTimestamp}`);
    if (maxTimestamp != null) query = query.concat(`&maxTimestamp=${maxTimestamp}`);
    if (filter != null) query = query.concat(`&filter=${filter}`);
    return query;
}

export const hpcFavoriteApp = (name: string, version: string) =>
    `/hpc/apps/favorites/${encodeURIComponent(name)}/${encodeURIComponent(version)}`;

export const hpcFavorites = (itemsPerPage: number, pageNumber: number) =>
    `/hpc/apps/favorites?itemsPerPage=${itemsPerPage}&page=${pageNumber}`;

export const hpcApplicationsQuery = (page: number, itemsPerPage: number) =>
    `/hpc/apps?page=${page}&itemsPerPage=${itemsPerPage}`;

interface HPCApplicationsSearchQuery {
    query: string,
    page: number,
    itemsPerPage: number
}

export const hpcApplicationsSearchQuery = ({query, page, itemsPerPage}: HPCApplicationsSearchQuery): string =>
    `/hpc/apps/search?query=${encodeURIComponent(query)}&page=${page}&itemsPerPage=${itemsPerPage}`;

export const hpcApplicationsTagSearchQuery = ({query, page, itemsPerPage}: HPCApplicationsSearchQuery): string =>
    `/hpc/apps/searchTags?query=${encodeURIComponent(query)}&page=${page}&itemsPerPage=${itemsPerPage}`;

export const cancelJobQuery = `hpc/jobs`;

export const cancelJobDialog = ({jobId, onConfirm, jobCount = 1}: { jobCount?: number, jobId: string, onConfirm: () => void }): void =>
    addStandardDialog({
        title: `Cancel job${jobCount > 1 ? "s" : ""}?`,
        message: jobCount === 1 ? `Cancel job: ${jobId}?` : "Cancel jobs?",
        cancelText: "No",
        confirmText: `Cancel job${jobCount > 1 ? "s" : ""}`,
        onConfirm
    });

export const cancelJob = (cloud: Cloud, jobId: string): Promise<{ request: XMLHttpRequest, response: void }> =>
    cloud.delete(cancelJobQuery, {jobId});

interface FavoriteApplicationFromPage<T> {
    name: string
    version: string
    page: Page<{ metadata: ApplicationMetadata, favorite: boolean } & T>
    cloud: Cloud
}

/**
 * Favorites an application.
 * @param {Application} Application the application to be favorited
 * @param {Cloud} cloud The cloud instance for requests
 */
export async function favoriteApplicationFromPage<T>({name, version, page, cloud}: FavoriteApplicationFromPage<T>): Promise<Page<T>> {
    const a = page.items.find(it => it.metadata.name === name && it.metadata.version === version)!;
    try {
        await cloud.post(hpcFavoriteApp(name, version));
        a.favorite = !a.favorite;
    } catch (e) {
        snackbarStore.addFailure(errorMessageOrDefault(e, `An error ocurred favoriting ${name}`));
    }
    return page;
}


interface StringMap {
    [k: string]: string;
}

interface AllowedParameterKey {
    name: string;
    type: ParameterTypes;
}

interface ExtractParameters {
    nameToValue: StringMap;
    allowedParameterKeys: AllowedParameterKey[];
    siteVersion: number;
}

export const findKnownParameterValues = ({nameToValue, allowedParameterKeys, siteVersion}: ExtractParameters): StringMap => {
    const extractedParameters = {};
    if (siteVersion === 1) {
        allowedParameterKeys.forEach(({name, type}) => {
            if (nameToValue[name] !== undefined) {
                if (typeMatchesValue(type, nameToValue[name])) {
                    extractedParameters[name] = nameToValue[name];
                }
            }
        });
    }
    return extractedParameters;
};

export const isFileOrDirectoryParam = ({type}: { type: string }) => type === "input_file" || type === "input_directory";

const typeMatchesValue = (type: ParameterTypes, parameter: string): boolean => {
    switch (type) {
        case ParameterTypes.Boolean:
            return parameter === "Yes" || parameter === "No" || parameter === "";
        case ParameterTypes.Integer:
            return parseInt(parameter, 10) % 1 === 0;
        case ParameterTypes.FloatingPoint:
            return typeof parseFloat(parameter) === "number";
        case ParameterTypes.Text:
        case ParameterTypes.InputDirectory:
        case ParameterTypes.InputFile:
        case ParameterTypes.SharedFileSystem:
        case ParameterTypes.Peer:
            return typeof parameter === "string";

    }
};

interface ExtractedParameters {
    [key: string]: string | number | boolean |
        { source: string, destination: string; } |
        { fileSystemId: string } |
        { jobId: string }
}

export type ParameterValues = Map<string, React.RefObject<HTMLInputElement | HTMLSelectElement>>;

interface ExtractParametersFromMap {
    map: ParameterValues;
    appParameters: ApplicationParameter[];
    cloud: Cloud;
}

export function extractValuesFromWidgets({map, appParameters, cloud}: ExtractParametersFromMap): ExtractedParameters {
    const extracted: ExtractedParameters = {};
    map.forEach(({current}, key) => {
        const parameter = appParameters.find(it => it.name === key);
        console.log(parameter, current);
        if (!current) return;
        if (!current.value || !current.checkValidity()) return;
        if (!parameter) return;
        switch (parameter.type) {
            case ParameterTypes.InputDirectory:
            case ParameterTypes.InputFile:
                const expandedValue = expandHomeFolder(current.value, cloud.homeFolder);
                extracted[key] = {
                    source: expandedValue,
                    destination: removeTrailingSlash(expandedValue).split("/").pop()!
                };
                return;
            case ParameterTypes.Boolean:
                switch (current.value) {
                    case "Yes":
                        extracted[key] = true;
                        return;
                    case "No":
                        extracted[key] = false;
                        return;
                    default:
                        return;
                }
            case ParameterTypes.Integer:
                extracted[key] = parseInt(current.value, 10);
                return;
            case ParameterTypes.FloatingPoint:
                extracted[key] = parseFloat(current.value);
                return;
            case ParameterTypes.Text:
                extracted[key] = current.value;
                return;
            case ParameterTypes.SharedFileSystem:
                console.log(current.value);
                extracted[key] = {fileSystemId: current.value};
                return;
            case ParameterTypes.Peer:
                extracted[key] = {jobId: current.value};
                return;
        }
    });
    return extracted;
}

export const inCancelableState = (state: AppState) =>
    state === AppState.VALIDATED ||
    state === AppState.PREPARED ||
    state === AppState.SCHEDULED ||
    state === AppState.RUNNING;


export function validateOptionalFields(
    invocation: ApplicationInvocationDescription,
    parameters: ParameterValues
): boolean {
    const optionalErrors = [] as string[];
    const optionalParams = invocation.parameters.filter(it => it.optional && it.visible).map(it =>
        ({name: it.name, title: it.title})
    );
    optionalParams.forEach(it => {
        const param = parameters.get(it.name)!;
        if (!param.current!.checkValidity()) optionalErrors.push(it.title);
    });

    if (optionalErrors.length > 0) {
        snackbarStore.addFailure(
            `Invalid values for ${optionalErrors.slice(0, 3).join(", ")}
                    ${optionalErrors.length > 3 ? `and ${optionalErrors.length - 3} others` : ""}`,
            5000
        );
        return false;
    }

    return true;
}

export function checkForMissingParameters(
    parameters: ExtractedParameters,
    invocation: ApplicationInvocationDescription
): boolean {
    const requiredParams = invocation.parameters.filter(it => !it.optional);
    const missingParameters: string[] = [];
    requiredParams.forEach(rParam => {
        const parameterValue = parameters[rParam.name];
        // Number, string, boolean 
        if (!parameterValue) {
            missingParameters.push(rParam.title);
        } else if (rParam.type === ParameterTypes.InputDirectory || rParam.type === ParameterTypes.InputFile) {
            if (!parameterValue["source"]) {
                missingParameters.push(rParam.title);
            }
        } else if (rParam.type === ParameterTypes.SharedFileSystem) {
            if (!parameterValue["fileSystemId"]) {
                missingParameters.push(rParam.title);
            }
        }
    });

    // Check missing values for required input fields.
    if (missingParameters.length > 0) {
        snackbarStore.addFailure(
            `Missing values for ${missingParameters.slice(0, 3).join(", ")} 
                ${missingParameters.length > 3 ? `and ${missingParameters.length - 3} others.` : ``}`,
            5000
        );
        return false;
    }
    return true;
}
