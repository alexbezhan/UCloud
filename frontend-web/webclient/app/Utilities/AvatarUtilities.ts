export const saveAvatarQuery = "/avatar/update";
export const findAvatarQuery = "/avatar/find";
