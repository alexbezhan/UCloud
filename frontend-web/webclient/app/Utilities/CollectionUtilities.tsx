export function removeEntry<T>(arr: T[], index: number): T[] {
    return arr.slice(0, index).concat(arr.slice(index + 1));
}

export function setsDiffer<T>(s1: Set<T>, s2: Set<T>): boolean {
    if (s1.size !== s2.size) return true;
    let differ = false;
    s1.forEach(it => {
        if (!s2.has(it)) differ = true;
    });
    return differ;
}

type PrimitiveDataTypes = string | number | boolean;

/**
 * @param set
 * @param entry
 */
export function addEntryIfNotPresent(set: Set<PrimitiveDataTypes>, entry: PrimitiveDataTypes): boolean {
    const size = set.size;
    set.add(entry);
    return size !== set.size;
}

export function groupBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T[]> {
    const result: Record<string, T[]> = {};
    items.forEach(item => {
        const key = keySelector(item);
        const newList = result[key] ?? [];
        newList.push(item);
        result[key] = newList;
    });
    return result;
}

export function associateBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T> {
    const result: Record<string, T> = {};
    items.forEach(item => {
        const key = keySelector(item);
        result[key] = item;
    });
    return result;
}

export function takeLast<T>(items: T[], numberOfItems: number): T[] {
    return items.slice(Math.max(0, items.length - numberOfItems));
}
