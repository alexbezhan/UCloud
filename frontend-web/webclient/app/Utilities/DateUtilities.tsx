import {format} from "date-fns/esm";

// Could potentially cause issues with time if user is outside CEST
export const dateToString = (date: number) => format(date, "HH:mm dd/MM/yyyy");
