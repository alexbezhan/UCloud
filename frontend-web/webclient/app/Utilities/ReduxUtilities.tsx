import activity from "Activity/Redux/ActivityReducer";
import * as AppRedux from "Applications/Redux";
import analyses from "Applications/Redux/AnalysesReducer";
import applications from "Applications/Redux/BrowseReducer";
import detailedApplicationSearch from "Applications/Redux/DetailedApplicationSearchReducer";
import * as TaskRedux from "BackgroundTasks/redux";
import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import detailedFileSearch from "Files/Redux/DetailedFileSearchReducer";
import fileInfo from "Files/Redux/FileInfoReducer";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import uploader from "Uploader/Redux/UploaderReducer";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    activity,
    dashboard,
    analyses,
    applications,
    header,
    status,
    sidebar,
    uploader,
    notifications,
    simpleSearch,
    detailedFileSearch,
    detailedApplicationSearch,
    fileInfo,
    hookStore,
    ...AppRedux.reducers,
    avatar,
    loading,
    tasks: TaskRedux.reducer,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
