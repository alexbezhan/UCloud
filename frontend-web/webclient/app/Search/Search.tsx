import {AdvancedSearchRequest as AppSearchRequest, DetailedApplicationSearchReduxState} from "Applications";
import {ApplicationCard} from "Applications/Card";
import DetailedApplicationSearch from "Applications/DetailedApplicationSearch";
import {setAppQuery} from "Applications/Redux/DetailedApplicationSearchActions";
import {Client} from "Authentication/HttpClientInstance";
import {emptyPage, HeaderSearchType, ReduxObject} from "DefaultObjects";
import {AdvancedSearchRequest, DetailedFileSearchReduxState, FileType} from "Files";
import DetailedFileSearch from "Files/DetailedFileSearch";
import {EmbeddedFileTable} from "Files/FileTable";
import {setFilename, toggleFilesSearchHidden} from "Files/Redux/DetailedFileSearchActions";
import {MainContainer} from "MainContainer/MainContainer";
import {setPrioritizedSearch, setRefreshFunction} from "Navigation/Redux/HeaderActions";
import {setActivePage} from "Navigation/Redux/StatusActions";
import * as Pagination from "Pagination";
import * as React from "react";
import {connect} from "react-redux";
import {Dispatch} from "redux";
import {Box, SelectableText, SelectableTextWrapper} from "ui-components";
import {GridCardGroup} from "ui-components/Grid";
import Hide from "ui-components/Hide";
import {SidebarPages} from "ui-components/Sidebar";
import {Spacer} from "ui-components/Spacer";
import {favoriteApplicationFromPage} from "Utilities/ApplicationUtilities";
import {searchPage} from "Utilities/SearchUtilities";
import {getQueryParamOrElse} from "Utilities/URIUtilities";
import {prettierString} from "UtilityFunctions";
import {SearchProps, SimpleSearchOperations, SimpleSearchStateProps} from ".";
import * as SSActions from "./Redux/SearchActions";

function Search(props: SearchProps) {
    React.useEffect(() => {
        props.toggleAdvancedSearch();
        props.setActivePage();
        const q = query(props);
        props.setSearch(q);
        props.setPrioritizedSearch(props.match.params.priority as HeaderSearchType);
        props.setRefresh(() => fetchAll());
        fetchAll();
        return () => {
            props.toggleAdvancedSearch();
            props.clear();
            props.setRefresh();
        };
    }, []);

    React.useEffect(() => {
        props.setPrioritizedSearch(props.match.params.priority as HeaderSearchType);
    }, [props.match.params.priority]);

    React.useEffect(() => {
        props.setSearch(props.search);
        fetchAll();
    }, [query(props)]);

    const setPath = (text: string) => props.history.push(searchPage(text.toLocaleLowerCase(), props.search));

    function fetchAll(itemsPerPage?: number) {
        props.searchFiles(fileSearchBody(
            props.fileSearch,
            props.search,
            itemsPerPage || props.files.itemsPerPage,
            props.files.pageNumber
        ));
        props.searchApplications(applicationSearchBody(
            props.applicationSearch,
            props.search,
            itemsPerPage || props.applications.itemsPerPage,
            props.applications.pageNumber
        ));
        props.history.push(searchPage(props.match.params.priority, props.search));
    }

    const refreshFiles = () => props.searchFiles(fileSearchBody(
        props.fileSearch,
        props.search,
        props.files.itemsPerPage,
        props.files.pageNumber
    ));

    const {files, applications, applicationsLoading} = props;

    const Tab = ({searchType}: {searchType: HeaderSearchType}): JSX.Element => (
        <SelectableText
            cursor="pointer"
            fontSize={3}
            onClick={() => setPath(searchType)}
            selected={priority === searchType}
            mr="1em"
        >
            {prettierString(searchType)}
        </SelectableText>
    );

    const allowedSearchTypes: HeaderSearchType[] = ["files", "applications"];

    let main: React.ReactNode = null;
    const {priority} = props.match.params;
    const entriesPerPage = (

        <Box my="8px">
            <Spacer
                left={null}
                right={(
                    <Pagination.EntriesPerPageSelector
                        onChange={itemsPerPage => fetchAll(itemsPerPage)}
                        content={`${prettierString(priority)} per page`}
                        entriesPerPage={
                            priority === "files" ? props.files.itemsPerPage :
                                props.applications.itemsPerPage
                        }
                    />
                )}
            />
        </Box>
    );
    if (priority === "files") {
        main = (
            <>
                <Hide xxl xl lg>
                    <DetailedFileSearch cantHide />
                    {entriesPerPage}
                </Hide>
                <EmbeddedFileTable
                    onPageChanged={page => props.searchFiles(
                        fileSearchBody(props.fileSearch, props.search, props.files.itemsPerPage, page)
                    )}
                    page={files ? files : emptyPage}
                    onReloadRequested={refreshFiles}
                    includeVirtualFolders={false}
                />
            </>
        );
    } else if (priority === "applications") {
        main = (
            <>
                <Hide xxl xl lg>
                    <DetailedApplicationSearch />
                    {entriesPerPage}
                </Hide>
                <Pagination.List
                    loading={applicationsLoading}
                    pageRenderer={({items}) => (
                        <GridCardGroup>
                            {items.map(app => (
                                <ApplicationCard
                                    onFavorite={async () => props.setApplicationsPage(await favoriteApplicationFromPage({
                                        name: app.metadata.name,
                                        version: app.metadata.version,
                                        page: props.applications,
                                        client: Client
                                    }))}
                                    key={`${app.metadata.name}${app.metadata.version}`}
                                    app={app}
                                    isFavorite={app.favorite}
                                    tags={app.tags}
                                />))}
                        </GridCardGroup>
                    )}
                    page={applications}
                    onPageChanged={pageNumber => props.searchApplications(
                        applicationSearchBody(
                            props.applicationSearch,
                            props.search,
                            props.applications.itemsPerPage,
                            pageNumber
                        ))
                    }
                />
            </>
        );
    }

    return (
        <MainContainer
            header={(
                <React.Fragment>
                    <SelectableTextWrapper>
                        {allowedSearchTypes.map((pane, index) => <Tab searchType={pane} key={index} />)}
                    </SelectableTextWrapper>
                    <Hide md sm xs>
                        <Spacer left={null} right={entriesPerPage} />
                    </Hide>
                </React.Fragment>
            )}
            main={main}
        />
    );
}

const mapDispatchToProps = (dispatch: Dispatch): SimpleSearchOperations => ({
    setFilesLoading: loading => dispatch(SSActions.setFilesLoading(loading)),
    setApplicationsLoading: loading => dispatch(SSActions.setApplicationsLoading(loading)),
    clear: () => dispatch(SSActions.receiveFiles(emptyPage)),
    searchFiles: async body => {
        dispatch(SSActions.setFilesLoading(true));
        dispatch(await SSActions.searchFiles(body));
        dispatch(setFilename(body.fileName || ""));
    },
    searchApplications: async body => {
        dispatch(SSActions.setApplicationsLoading(true));
        dispatch(await SSActions.searchApplications(body));
        dispatch(setAppQuery(body.query || ""));
    },
    setFilesPage: page => dispatch(SSActions.receiveFiles(page)),
    setApplicationsPage: page => dispatch(SSActions.receiveApplications(page)),
    setSearch: search => dispatch(SSActions.setSearch(search)),
    setPrioritizedSearch: sT => dispatch(setPrioritizedSearch(sT)),
    toggleAdvancedSearch: () => dispatch(toggleFilesSearchHidden()),
    setActivePage: () => dispatch(setActivePage(SidebarPages.None)),
    setRefresh: refresh => dispatch(setRefreshFunction(refresh)),
});

const mapStateToProps = ({
    simpleSearch,
    detailedFileSearch,
    detailedApplicationSearch
}: ReduxObject): SimpleSearchStateProps & {favFilesCount: number, favAppCount: number} => ({
    ...simpleSearch,
    favFilesCount: simpleSearch.files.items.filter(it => it.favorited).length,
    favAppCount: simpleSearch.applications.items.filter(it => it.favorite).length,
    fileSearch: detailedFileSearch,
    applicationSearch: detailedApplicationSearch
});

export default connect(mapStateToProps, mapDispatchToProps)(Search);

export function fileSearchBody(
    fileSearch: DetailedFileSearchReduxState,
    fileName: string,
    itemsPerPage: number,
    page: number
): AdvancedSearchRequest {
    const fileTypes: [FileType?, FileType?] = [];
    if (fileSearch.allowFiles) fileTypes.push("FILE");
    if (fileSearch.allowFolders) fileTypes.push("DIRECTORY");
    const modifiedAt = {
        after: !!fileSearch.modifiedAfter ? fileSearch.modifiedAfter.valueOf() : undefined,
        before: !!fileSearch.modifiedBefore ? fileSearch.modifiedBefore.valueOf() : undefined,
    };

    return {
        fileName,
        extensions: [...fileSearch.extensions],
        fileTypes,
        modifiedAt: typeof modifiedAt.after === "number" ||
            typeof modifiedAt.before === "number" ? modifiedAt : undefined,
        includeShares: fileSearch.includeShares,
        itemsPerPage,
        page
    };
}

export function applicationSearchBody(
    body: DetailedApplicationSearchReduxState,
    appName: string,
    itemsPerPage: number,
    page: number
): AppSearchRequest {
    const {tags, showAllVersions} = body;
    return {
        query: appName,
        tags: tags.size > 0 ? [...tags] : undefined,
        showAllVersions,
        itemsPerPage,
        page
    };
}

function query(props: SearchProps): string {
    return queryFromProps(props);
}

function queryFromProps(p: SearchProps): string {
    return getQueryParamOrElse(p, "query", "");
}
