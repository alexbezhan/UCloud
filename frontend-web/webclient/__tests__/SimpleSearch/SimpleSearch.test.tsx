import {configure} from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import {createBrowserHistory} from "history";
import * as React from "react";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router";
import {create} from "react-test-renderer";
import {ThemeProvider} from "styled-components";
import Search from "../../app/Search/Search";
import theme from "../../app/ui-components/theme";
import {store} from "../../app/Utilities/ReduxUtilities";

configure({adapter: new Adapter()});

test("Search mount", () => {
    expect(
        create(
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <MemoryRouter>
                        <Search
                            match={{
                                isExact: false,
                                params: {priority: "FILES"},
                                url: "",
                                path: ""
                            }}
                            history={createBrowserHistory()}
                            location={{search: ""}}
                        />
                    </MemoryRouter>
                </ThemeProvider>
            </Provider>
        )
    ).toMatchSnapshot();
});
