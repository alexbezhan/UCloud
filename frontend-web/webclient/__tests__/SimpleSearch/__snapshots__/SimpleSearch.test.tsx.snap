// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search mount 1`] = `
.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  margin-left: 190px;
  padding-top: 48px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  background-color: #fff;
}

.c6 {
  margin-left: auto;
}

.c7 {
  margin-top: 8px;
  margin-bottom: 8px;
}

.c12 {
  padding-top: 96px;
  padding-right: 14px;
}

.c11 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: inherit;
  margin-left: .7em;
}

.c9 {
  cursor: pointer;
}

.c10 {
  cursor: inherit;
}

.c8 {
  position: relative;
  display: inline-block;
}

.c8:hover > div {
  display: block;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 2px solid #c9d3df;
  cursor: pointer;
}

.c3 {
  cursor: pointer;
  font-size: 20px;
  margin-right: 1em;
}

.c1 {
  padding-top: 14px;
  padding-bottom: 14px;
  padding-left: 204px;
  padding-right: 14px;
  width: 100%;
  height: 96px;
  background-color: #fff;
  position: absolute;
  top: 48px;
  left: 0;
  position: fixed;
  z-index: 50;
}

@media screen and (max-width:511px) {
  .c4 {
    display: none;
  }
}

@media screen and (min-width:512px) and (max-width:639px) {
  .c4 {
    display: none;
  }
}

@media screen and (min-width:640px) and (max-width:767px) {
  .c4 {
    display: none;
  }
}

@media screen and (min-width:768px) and (max-width:1023px) {
  .c13 {
    display: none;
  }
}

@media screen and (min-width:1024px) and (max-width:1279px) {
  .c13 {
    display: none;
  }
}

@media screen and (min-width:1280px) {
  .c13 {
    display: none;
  }
}

<div
  className="c0"
>
  <div
    className="sc-AxgMl c1"
    height={96}
    width={1}
  >
    <div
      className="c2"
    >
      <div
        className="c3"
        cursor="pointer"
        fontSize={3}
        onClick={[Function]}
        selected={false}
      >
        Files
      </div>
      <div
        className="c3"
        cursor="pointer"
        fontSize={3}
        onClick={[Function]}
        selected={false}
      >
        Applications
      </div>
    </div>
    <div
      className="c4"
    >
      <div
        className="c5"
      >
        <div
          className="c6"
        />
        <div
          className="c7"
        >
          <div
            className="c5"
          >
            <div
              className="c6"
            />
            <div
              className="c8"
              data-tag="dropdown"
            >
              <span
                className="c9"
                cursor="pointer"
                onClick={[Function]}
              >
                <span
                  className="c10"
                  cursor="inherit"
                >
                   
                  Files per page 25
                </span>
                <svg
                  className="c11"
                  clipRule="evenodd"
                  cursor="inherit"
                  fill="currentcolor"
                  fillRule="evenodd"
                  height=".7em"
                  ml=".7em"
                  viewBox="0 0 24 20"
                  width=".7em"
                >
                  <path
                    d="M11.998 19.999c-.598 0-1.028-.61-1.028-.61L.365 2.268C-.453 1.02.153 0 1.703 0h20.593c1.554 0 2.156 1.02 1.34 2.27l-10.61 17.121c0-.002-.43.608-1.028.608z"
                    fillRule="nonzero"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c12"
  >
    <div
      className="c13"
    />
  </div>
</div>
`;
